projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,383 @@
+print("Updated builds are identical:\t{:.3f}%"\
+      .format(pct(original_builds_same.get(True), sum(original_builds_same.values()))))
+print("Updated builds are different:\t{:.3f}%"\
+      .format(pct(original_builds_same.get(False), sum(original_builds_same.values()))))
+```
+    Updated builds are identical:	16.472%
+    Updated builds are different:	83.528%
+
+
+The data shows that 83.52% of the 1.31% dupes are updating from different builds. The 0.22% of all `update` pings that are the same client updating to the same build from the same build are, at present, unexplained (but in small enough quantities we can ignore for the moment).
+
+The `update` pings with the same previous build information may be coming from the same profile, copied and then used with different versions of Firefox. Depending on when the browser is started with a specific copied profile, the downloaded *update* blob might be different (more recent), thus resulting in an `update` with `reason = success` being sent with the same *previous build* information but with different *current build* information.
+
+## Validate the submission delay
+
+### How long until we receive the ping after it's created?
+
+
+```python
+delays = deduped_subset.map(lambda p: calculate_submission_delay(p))
+```
+
+```python
+setup_plot("'update' ('success') ping submission delay CDF",
+           MAX_DELAY_S / HOUR_IN_S, area_border_x=1.0)
+
+plot_cdf(delays\
+         .map(lambda d: d / HOUR_IN_S if d < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+         .collect(), label="CDF", linestyle="solid")
+
+plt.show()
+```
+
+
+![png](images/output_23_0.png)
+
+
+Almost 95% of the `update` pings with `reason = success` are submitted within an hour from the ping being created. Since we know that this ping is created as soon as the [update is applied](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/update-ping.html#payload-reason) we can claim that we receive 95% of these pings within an hour from the update being applied. 
+
+## Make sure that the volume of incoming pings is reasonable
+Check if the volume of `update` pings with `reason = ready` matches with the volume of pings with `reason = success`. For each ping with `reason = ready`, find the matching ping with `reason = success`.
+
+We are considering the data within a very narrow window of time: we could see `reason = success` pings from users that sent a `reason = ready` ping before the 3rd of September and `reason = ready` pings from users that have sent us a `reason = success` after the 9th of September. Filter these edge cases out by inspecting the `previousBuildId` and `targetBuildId`.
+
+
+```python
+filtered_ready = ping_ready.filter(lambda p: p.get("payload/targetBuildId") < "{}000000".format(MAX_DATE))
+filtered_success = ping_success.filter(lambda p: p.get("payload/previousBuildId") >= "{}999999".format(MIN_DATE))
+```
+Use the filtered RDDs to match betweek the different ping reasons.
```

Comment:
I wish "betweek" was a word. It sounds awesome.

Reply:
Team motto 2018? :)